### ISSUE TITLE
Using a Trained Model

### ISSUE THREAD

**Issue by MiracAydin1** [usnistgov/alignn#50]
Dear All,

I successfully trained a model with ALIGNN. There is "pretrained.py" script that import some models according to datasets.

My question is how I can use my trained model to predict a structure? Do I have to modify the "pretrained.py" script to import my model or can I do something like this:

`python alignn/pretrained.py --model_name <my_output_pt_file> --file_format poscar --file_path /path/to/sample_file`

Thanks for the help.

**Comment by bdecost** [comment]
It's probably easier to write your own script based off this one. This script grabs model checkpoint zip archives from figshare, instantiates  [an ALIGNN model with default hyperparameters here](https://github.com/usnistgov/alignn/blob/dba50f78f46385e53e901db8956a568febe82444/alignn/pretrained.py#L197) (except for `output_features`), and loads a single structure from the disk to run through the network. (Refer to the [default configuration here](https://github.com/usnistgov/alignn/blob/dba50f78f46385e53e901db8956a568febe82444/alignn/models/alignn.py#L22))

You should make sure to instantiate your model using your training config if you've tuned any of these hyperparameters. Then you can just load your parameters from your model checkpoint file [like this](https://github.com/usnistgov/alignn/blob/dba50f78f46385e53e901db8956a568febe82444/alignn/pretrained.py#L202), since you don't need to download and decompress from figshare.

then you probably want a loop around all your structures, building graphs and running them through the model [like this](https://github.com/usnistgov/alignn/blob/dba50f78f46385e53e901db8956a568febe82444/alignn/pretrained.py#L209)

If you have a pretty large dataset or you want to do repeated evaluations or something, it's probably worth it to write a dataloader and adapt some code in [alignn/train.py](https://github.com/usnistgov/alignn/blob/main/alignn/train.py) so you can take advantage of parallelism. Organizing that code to make this sort of thing easier has been on my list of things to get to for a while...

**Comment by sgbaird** [comment]
@bdecost I'd be very interested in using ALIGNN to predict properties for many structures. For now, I think we'll need to switch to CGCNN due to time/implementation constraints, but if ALIGNN gets a parallel prediction interface in the future, definitely looking forward to it 🙂

**Comment by bdecost** [comment]
@sgbaird thanks for the feedback, I agree it's an important feature request. Can you say a bit more about your use case, and what you think would make for a convenient interface / example? 

If you have a large static dataset of structures I think it's mostly a matter of setting up a [StructureDataset](https://github.com/usnistgov/jarvis/blob/217eb3719f9ffce8f2bde9ae73f3961e42660217/jarvis/core/graphs.py#L522) with your data, and cleaning up the prediction interface. That should be enough for batch single GPU prediction, not sure how much extra work it might be for distributed GPU prediction.

currently the jarvis StructureDataset expects a pandas.DataFrame following the format from `jarvis.db.figshare.data`, including a target column -- that could be modified to allow a missing target column, or maybe just pass NaN values

If you're generating lots of structures dynamically then there are other considerations, and the overhead of precomputing graphs might need to be addressed. I usually precompute and cache graphs to disk for repeated training runs. Maybe DataLoader prefetching will be enough to hide this overhead, I'm not sure.

I can tackle a lot of this core codebase stuff, but right now I don't think I can realistically promise enough bandwidth to not risk being a bottleneck on your project 😬  -- so I don't want to dissuade you from switching to CGCNN

**Comment by knc6** [comment]
@bdecost @sgbaird 
The following module `get_multiple_predictions` might be useful:

https://github.com/usnistgov/alignn/blob/develop/alignn/pretrained.py#L245

**Comment by bdecost** [comment]
@knc6 thanks, I didn't realize you had this ready to go already

**Comment by sgbaird** [comment]
That seemed to do the trick
https://colab.research.google.com/github/sparks-baird/alignn/blob/main/Training_ALIGNN_model_example.ipynb#scrollTo=v62Vzv2_2M2s

**Comment by sgbaird** [comment]
Didn't realize it was already released:
https://github.com/usnistgov/alignn/releases/tag/v2022.01.25

**Comment by knc6** [comment]
Yes, we thought it might be useful. Please feel free to let us know if you encounter any issue further.
